rust/libnewsboat/src/utils.rs

```diff
@@ -228,6 +231,25 @@ pub fn is_valid_podcast_type(mimetype: &str) -> bool {
     matches || found
 }
 
+pub fn escape_url(rs_str: String) -> String {
+    define_encode_set! {
+        pub URL_ENCODE_SET = [SIMPLE_ENCODE_SET] | {' ','!','#','$','&','\'','(',')','*','+',',','/',':',';','=','?','@','[',']'}
+    }
+    percent_encode(rs_str.as_bytes(),URL_ENCODE_SET).to_string()
+}
+
+pub fn unescape_url(rs_str: String) -> String {
+    let result = percent_decode(rs_str.as_bytes());
+    let result = result.decode_utf8();
+    if result.is_err() {
+        log!(Level::Error,
+                    &format!("percent_decode failed to escape url {}", rs_str));
+        panic!("Escaping url Failed");
```

Comment:
C++ implementation threw an exception here, and handled it e.g. in [src/feedhqapi.cpp](https://github.com/newsboat/newsboat/blob/9daed679891bfb52b2c115e80cc6f1418c118190/src/feedhqapi.cpp#L191-L200) (The Old Reader API does the same, while Inoreader doesn't). However, this Rust implementation will crash the program without a chance to recover. Since the error is probably recoverable, I don't think the panic is the right approach here.

This is the first time we're porting something that can throw an exception, so I don't have a ready solution to point to. There is no way to throw an exception over FFI (from Rust to C++), so we'll have to emulate it. I think it makes sense to:
1. return `Option<String`. This is the Rust way of dealing with exceptional situations;
2. `rs_unescape_url` should return NULL if `unescape_url` returned `None`;
3. on C++ side, `unescape_url` should throw the old exception if `rs_unescape_url` returned NULL, otherwise return `RustString`.

This way, we get to keep the old C++ interface on the C++ side (so no refactoring needed), and we also get a nice-to-use Rust version that integrates well with how Rust handles errors.

Do you see any flaws with this? Can you suggest something better?

Reply:
How would you feel about moving `escape_url()`s functionality to `rs_unescape_url()`? 
`unescape_url()` is really just a wrapper around `percent_decode()`, and as soon as utils.cpp stops existing it will have no reason to exist. The same could be said for several other functions in utils.rs.
That way we would just skip step 1 and just do 2 and 3.